routes/util.go

```diff
@@ -24,3 +33,106 @@ func (d *deps) isIgnored(name string) bool {
 
 	return false
 }
+
+type repository struct {
+	Name        string
+	Category    string
+	Path        string
+	Slug        string
+	Description string
+	LastCommit  string
+}
+
+type entry struct {
+	Name         string
+	Repositories []*repository
+}
+
+type entries struct {
+	Children []*entry
+	c        map[string]*entry
+}
+
+func (ent *entries) Add(r repository) {
+	if r.Category == "" {
+		ent.Children = append(ent.Children, &entry{
+			Name:         r.Name,
+			Repositories: []*repository{&r},
+		})
+		return
+	}
+	t, ok := ent.c[r.Category]
+	if !ok {
+		t := &entry{
+			Name:         r.Category,
+			Repositories: []*repository{&r},
+		}
+		ent.c[r.Category] = t
+		ent.Children = append(ent.Children, t)
+		return
+	}
+	t.Repositories = append(t.Repositories, &r)
+}
+
+func (d *deps) getAllRepos() (*entries, error) {
+	entries := &entries{
+		Children: []*entry{},
+		c:        map[string]*entry{},
+	}
+	max := strings.Count(d.c.Repo.ScanPath, string(os.PathSeparator)) + 2
+
+	err := filepath.WalkDir(d.c.Repo.ScanPath, func(path string, de fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if de.IsDir() {
+			// Check if we've exceeded our recursion depth
+			if strings.Count(path, string(os.PathSeparator)) > max {
+				return fs.SkipDir
+			}
+
+			if d.isIgnored(path) {
+				return fs.SkipDir
+			}
+
+			// A bare repo should always have at least a HEAD file, if it
+			// doesn't we can continue recursing
+			if _, err := os.Lstat(filepath.Join(path, "HEAD")); err == nil {
+				repo, err := git.Open(path, "")
+				if err != nil {
+					log.Println(err)
+				} else {
+					relpath, _ := filepath.Rel(d.c.Repo.ScanPath, path)
```

Comment:
I'm ignoring the error here because the condition under which `filepath.Rel` could error cannot currently occur. `d.c.Repo.ScanPath` is an absolute path, and so is the path we get in `filepath.WalkDir` because we're walking from `d.c.Repo.ScanPath`